File: app/src/main/java/com/pr0gramm/app/ui/dialogs/NewTagDialogFragment.java
package com.pr0gramm.app.ui.dialogs;

import android.app.Dialog;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.widget.MultiAutoCompleteTextView;

import com.google.common.base.Splitter;
import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.R;
import com.pr0gramm.app.ui.DialogBuilder;
import com.pr0gramm.app.ui.TagInputView;
import com.pr0gramm.app.ui.base.BaseDialogFragment;

import java.util.List;

import butterknife.BindView;

/**
 */
public class NewTagDialogFragment extends BaseDialogFragment {
    @BindView(R.id.tag)
    MultiAutoCompleteTextView tagInput;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        return DialogBuilder.start(getActivity())
                .layout(R.layout.dialog_add_tags)
                .negative(R.string.cancel)
                .positive(R.string.dialog_action_add, this::onOkayClicked)
                .build();
    }

    @Override
    protected void onDialogViewCreated() {
        TagInputView.setup(tagInput);
    }

    private void onOkayClicked() {
        String text = tagInput.getText().toString();

        // split text into tags.
        Splitter splitter = Splitter.on(",").omitEmptyStrings().trimResults();
        List<String> tags = splitter.splitToList(text);

        // do nothing if the user had not typed any tags
        if (tags.isEmpty())
            return;

        // inform parent
        ((OnAddNewTagsListener) getParentFragment()).onAddNewTags(tags);
    }

    @Override
    protected void injectComponent(ActivityComponent activityComponent) {
        activityComponent.inject(this);
    }

    /**
     * The parent fragment must implement this interface.
     * It will be informed by this class if the user added tags.
     */
    public interface OnAddNewTagsListener {
        /**
         * Called when the dialog finishes with new tags.
         */
        void onAddNewTags(List<String> tags);
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/dialogs/NewTagDialogFragment.kt
package com.pr0gramm.app.ui.dialogs

import android.app.Dialog
import android.view.View
import androidx.core.view.isVisible
import com.pr0gramm.app.R
import com.pr0gramm.app.databinding.DialogAddTagsBinding
import com.pr0gramm.app.ui.TagSuggestionService
import com.pr0gramm.app.ui.base.ViewBindingDialogFragment
import com.pr0gramm.app.ui.dialog
import com.pr0gramm.app.util.AndroidUtility
import com.pr0gramm.app.util.addTextChangedListener
import com.pr0gramm.app.util.di.instance
import com.pr0gramm.app.util.hideSoftKeyboard

/**
 */
class NewTagDialogFragment :
        ViewBindingDialogFragment<DialogAddTagsBinding>("NewTagDialogFragment", DialogAddTagsBinding::inflate) {

    private val tagSuggestions: TagSuggestionService by instance()

    override fun onCreateDialog(contentView: View): Dialog {
        return dialog(this) {
            contentView(contentView)
            negative(R.string.cancel) { onCancelClicked() }
            positive(R.string.dialog_action_add) { onOkayClicked() }

            onShow {
                AndroidUtility.showSoftKeyboard(views.tagInput)
            }
        }
    }

    override fun onDialogViewCreated() {
        tagSuggestions.setupView(views.tagInput)

        views.tagInput.addTextChangedListener { text ->
            views.opinionHint.isVisible = tagSuggestions.containsQuestionableTag(text)
        }
    }

    private fun onOkayClicked() {
        val text = views.tagInput.text.toString()

        // split text into tags.
        val tags = text.split(',', '#').map { it.trim() }.filter { it.isNotEmpty() }

        // do nothing if the user had not typed any tags
        if (tags.isEmpty())
            return

        // inform parent
        (parentFragment as OnAddNewTagsListener).onNewTags(tags)

        hideSoftKeyboard()
    }

    private fun onCancelClicked() {
        hideSoftKeyboard()
    }

    /**
     * The parent fragment must implement this interface.
     * It will be informed by this class if the user added tags.
     */
    interface OnAddNewTagsListener {
        /**
         * Called when the dialog finishes with new tags.
         */
        fun onNewTags(tags: List<String>)
    }
}
